docs/custom_promise_types/cfengine.sh

```diff
@@ -0,0 +1,153 @@
+        promise_type)
+            request_promise_type="$value" ;;
+        promiser)
+            request_promiser="$value" ;;
+        attribute_*)
+            attribute_name=${key#"attribute_"}
+            if ! expr " $obligatory_attributes $optional_attributes " : ".* $attribute_name " >/dev/null; then
+                saw_unknown_attribute="yes"
+                unknown_attribute_names="$unknown_attribute_names, $attribute_name"
+            fi
+            eval "request_${key}=\$value" ;;
+        *)
+            saw_unknown_key="yes" ;;
+    esac
+}
+
+receive_request() {
+    # Read lines from input until empty line
+    # Call handle_input_line for each non-empty line
+    while IFS='$\n' read -r line; do
+        if [ "x$line" = "x" ] ; then
+            break
+        fi
+        handle_input_line "$line" # Parses a key=value pair
+    done
+}
+
+write_response() {
+    echo "operation=$request_operation"
+    echo "result=$response_result"
+    echo ""
+}
+
+operation_terminate() {
+    response_result="success"
+    type do_terminate >/dev/null && do_terminate
+    write_response
+    exit 0
+}
+
+operation_validate() {
+    response_result="valid"
+    if [ "$saw_unknown_attribute" != "no" -a "$all_attributes_are_valid" != "yes" ] ; then
+        echo "log_error=Unknown attribute/s: ${unknown_attribute_names#, }"
+        response_result="invalid"
+    fi
+
+    if [ ! -z "$obligatory_arrtibutes" ]; then
+        for attribute_name in $obligatory_attributes; do
+            varname="attribute_$obligatory_arrtibutes"
+            if [ -z "${!varname}" ]; then
+                echo "log_error=Attribute '$attribute_name' is missing or empty"
+                response_result="invalid"
+            fi
+        done
+    fi
+
+    type do_validate >/dev/null && do_validate
+    write_response
+}
+
+operation_evaluate() {
+    response_result="error" # it's responsibility of do_evaluate to override this
+    type do_evaluate >/dev/null && do_evaluate
+    write_response
+}
+
+operation_unknown() {
+    response_result="error"
+    echo "log_error=Promise module received unexpected operation: $request_operation"
+    write_response
+}
+
+perform_operation() {
+    case "$request_operation" in
+        validate_promise)
+            operation_validate ;;
+        evaluate_promise)
+            operation_evaluate ;;
+        terminate)
+            operation_terminate ;;
+        *)
+            operation_unknown ;;
+    esac
+}
+
+handle_request() {
+    reset_state         # 1. Reset global variables
+    receive_request     # 2. Receive / parse an operation from agent
+    perform_operation   # 3. Perform operation (validate, evaluate, terminate)
+}
+
+skip_header() {
+    # Skip until (and including) the first empty line
+    while IFS='$\n' read -r line; do
+        if [ "x$line" = "x" ] ; then
+            return;
+        fi
+    done
+}
+
+# Skip the protocol header given by agent:
+skip_header
```

Comment:
Should this really just always skip the header?

Reply:
Adding library support for getting the information from the header would be nice, but it's also quite safe to skip it now.

The header becomes important when we add more features / protocol versions, and we want to write modules which are compatible with both old and new agents.